postgresql: improve passthru.tests

wolfgangwalther opened the pull request:
`postgresql` and `postgresqlPackages`'s tests currently have the following problems:
- some of the nixosTests are missing as passthru tests for the postgresql package
- most extension's passthru tests always run with the default postgresql version, e.g. `postgresql_12_jit.pkgs.postgis.tests` actually tests the extension compiled again `postgresql_16`.

This PR fixes that - and more. The general approach is to provide a `passthru.override` function from each of the postgresql related nixosTests - and then use that to override the `postgresqlPackage` from the derivation/extension.

Furthermore, some tests (postgis and apache_datasketches) were simple enough to easily migrate them to `postgresqlTestExtension` - which doesn't need any VM, but just runs the tests via `postgresqlTestHook`. This is much more portable, as it should also work for darwin (do nixosTests work there, normally?) or could also work for `pkgsMusl.postgresql...` (didn't test, though).

Most importantly.. it would have allowed to run all those tests against the new PostgreSQL version, when we added that in #351253. Instead, they were all still running against the default version in that PR.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

wolfgangwalther committed:
nixosTests.postgresql-wal-receiver: avoid manual imports

Manually importing postgresql packages from the /pkgs/ folder or
manually importing the test from /nixos/tests/ in generic.nix is not
only ugly, but also forbidden should we ever move to pkgs/by-name.

We can achieve almost the same with a slightly different setup. We allow
overriding the postgresql package for the test via passthru.override, to
make sure that each postgresql_xx.tests.postgresql-wal-receiver is
properly teted with the right version.

wolfgangwalther committed:
nixosTests.postgresql-jit: avoid manual imports

Same reasoning as commit before.

wolfgangwalther committed:
postgresql: add passthru.tests.postgresql

Restructuring the nixosTests.postgresql test a little bit to allow
calling it with the specific versioned package from generic.nix.

wolfgangwalther committed:
postgresql: add passthru.tests.postgresql-tls-client-cert

Same reasoning as commit before.

wolfgangwalther committed:
postgresqlPackages.postgis: move nixosTests.postgis into package

Same reasoning as commit before.

wolfgangwalther committed:
nixosTests.postgresql-wal2json: avoid manual imports

Same reasoning as commit before.

wolfgangwalther committed:
postgresqlPackages.pgjwt: make passthru.tests work with correct package

Same reasoning as commit before.

wolfgangwalther committed:
postgresqlPackages.pgvecto-rs: make passthru.tests work with correct package

Same reasoning as commit before.

wolfgangwalther committed:
postgresqlPackages.tsja: make passthru.tests work with correct package

Same reasoning as commit before.

wolfgangwalther committed:
postgresqlPackages.apache_datasketches: move nixosTests.apache_datasketches into package

There is no need to fire up a whole VM just to run a two line test of
creating the extension. We can use postgresqlTestExtension for that.
This has the advantage that it runs with postgresqlTestHook, so without
a VM, making it more portable.

wolfgangwalther committed:
postgresqlPackages.timescaledb: make passthru.tests work with correct package

Same reasoning as commit before.

Ma27 reviewed (commented):
I like the change, thanks!

Just one question: what's the recommended way of running tests for all versions now?

Ma27 commented:
> I wonder whether we can somehow have an attribute somewhere, which would make it easy to do both at the same time?

I mean we have all information available for that.
I'm just not entirely sure where.

The easiest approach would be to add this to the respective files (i.e. `nixos/tests/postgresql.nix`), but this would be kinda awkward because we'd then have the `passthru` thing to create tests for a given version and all instantiated versions right next to it.

I don't know. I think I could also live with this approach though.

Ma27 commented:
Yes.
Maybe we want to have a small script in `maintainers/scripts` that lets you run tests for all postgresql versions and optionally only a subset of versions or a subset of tests?

Ma27 commented:
Hmm, it just occurred to me: are `passthru.tests` things even run by Hydra? IIRC this isn't the case and thus we need some kind of alternative exposure.

Given that we want a `nixos/tests/postgresql` subtree anyways, how about we add a `default.nix` in there that adds attributes for each test-case and for each version?

Then we can do `nix-build nixos/tests/postgresql` to run all tests.
And ecah `pkgs.postgresql` needs to pick the correct attributes from `pkgs.nixosTests` in `passthru.tests` (same for all extensions having test-cases).

I think Nextcloud does it that way already if you'd like to see it in the wild.

Does that sound reasonable? Are you interested in doing that? Otherwise, I'd be willing to take over soonish (at least not today).

Ma27 commented:
> The one thing that will still not be possible with that is to build all non-broken extensions for one (or all) postgresql version at once

I mean, it was the case already that you could only build extensions selectively or all, right?

> which might not always be in nixos/tests/postgresql, because they are defined in the package itself

So if we have a `default.nix` in `nixos/tests/postgresql.nix` that exposes all tests as attributes (also for each psotgresql version), then this is also exposed via `pkgs.nixosTests`. This means that we could just do `passthru.tests = { inherit (nixosTests.postgresql) postgresql_version postgresql_jit_version postgresql_wal_receiver_version ...; }`, right?

Ma27 commented:
> Yes. But doing nix-build -A postgresql.pkgs is really annoying, because of broken packages.

I fully agree here, I just don't have a really good idea right now.

> or are you thinking of nixosTests.postgresql.default and nixosTests.postgresql.wal-receiver etc. instead?

That'd be my suggestion, yes. May I ask what downsides you see with this?

> Do you want to run tests for all postgres versions when you build postgresql.tests?

No: my suggestion is to add `nixosTests.postgresql.pg_16` to `pkgs.postgresql_16.tests` basically.

> But I think the versioned attributes would still be one level below that.

Would be fine by me.

> Let's see what I come up with. The one thing that is still missing in my model is the extensions thing - but we can come to that later.

Yeah sure. I think the fact that the correct postgresql is tested in `postgresql.tests` and a subtree we can use for ci/OWNERS are two useful improvements already.

wolfgangwalther committed:
nixosTests.postgresql: move all postgresql related nixosTests into one folder

This makes it possible to run all those tests at once by building
nixosTests.postgresql and allow a simple entry to ci/OWNERS for all
tests.

wolfgangwalther committed:
nixosTests.pgjwt: fix test

This seems to have broken years ago, because "CREATE EXTENSION
pgcrypto;" etc. were added to the upstream file about 6 years ago.

wolfgangwalther committed:
postgresqlPackages.anonymizer: make passthru.tests work with correct package

Same reasoning as commit before.

wolfgangwalther committed:
postgresqlVersions: init

Allows building all PostgreSQL versions at once with:

  nix-build -A postgresqlVersions

Also makes it possible for nixosTests to target all PostgreSQL versions
without importing the postgresql folder across the whole repo.

wolfgangwalther committed:
nixosTests.postgresql: use a common pattern throughout all tests

Avoiding "with", using the same names and basic structure in each test.

Consistency is key!

wolfgangwalther committed:
nixosTests.postgresql: run nixfmt

Because with as many changes as in here anybody working on those test
files will have merge conflicts anyway.

Ma27 commented:
> This seems to take time and I doubt we can make it much faster right now. If nix (lix) was only evaluating in parallel...

You may want to use `nix-eval-jobs` then (or `nix-fast-build` - that already marries it together with `nom`), at the expense of more RAM (but I mean, a reason why I upgraded my laptop's RAM to 40GiB was to be able to evaluate all postgresql tests locally :see_no_evil: )

Thanks a lot!
I'll review it soonish, but it'll take a few days. I already did way more nixpkgs stuff this weekend than I intended.

Ma27 reviewed (commented):
THe merge conflicts and the note about the `package` argument need to be resolved. Other than that, it looks pretty good, thanks!

Ma27 commented:
I'm afraid we have another merge conflict :see_no_evil:
